File: mcspring-api/mcspring-base/src/main/java/in/kyle/mcspring/command/CommandController.java
package in.kyle.mcspring.command;

import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.command.CommandMap;
import org.bukkit.plugin.Plugin;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Controller;

import java.lang.reflect.Field;

import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;

@Controller
@RequiredArgsConstructor
class CommandController {
    
    private final CommandMap commandMap = getCommandMap();
    private final Plugin plugin;
    
    public void registerCommand(Command command) {
        commandMap.register(command.getLabel(), plugin.getName(), command);
    }
    
    @SneakyThrows
    private CommandMap getCommandMap() {
        Field bukkitCommandMap = Bukkit.getServer().getClass().getDeclaredField("commandMap");
        bukkitCommandMap.setAccessible(true);
        return (CommandMap) bukkitCommandMap.get(Bukkit.getServer());
    }
}

File: mcspring-api/mcspring-base/src/main/kotlin/in/kyle/mcspring/command/CommandController.kt
package `in`.kyle.mcspring.command

import org.bukkit.Bukkit
import org.bukkit.command.Command
import org.bukkit.command.CommandMap
import org.bukkit.plugin.Plugin
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean
import org.springframework.context.annotation.Bean
import org.springframework.stereotype.Controller

@Controller
@ConditionalOnBean(Plugin::class)
internal class CommandController(
        private val plugin: Plugin
) {
    private val commandMap = getCommandMap()

    fun registerCommand(command: Command) {
        commandMap.register(command.label, plugin.name, command)
    }

    private fun getCommandMap(): CommandMap {
        val bukkitCommandMap = Bukkit.getServer().javaClass.getDeclaredField("commandMap")
        bukkitCommandMap.isAccessible = true
        return bukkitCommandMap[Bukkit.getServer()] as CommandMap
    }

    @Bean
    fun commandMap(): CommandMap {
        return commandMap
    }
}
